rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,359 @@
+---
+authors: STeve Huang (xin.huang@goteleport.com)
+state: draft
+---
+
+# RFD 0209 - MCP Access
+
+## Required Approvers
+
+* Engineering: @r0mant
+* Product: @klizhentas
+
+## What
+
+Support zero-trust access for MCP servers.
+
+## Why
+
+Introduced in late 2024, Anthropic’s [Model Context Protocol
+(MCP)](https://modelcontextprotocol.io/introduction) is a widely adopted,
+open-source standard that enables language models to seamlessly interact with
+external tools and data, enhancing their contextual capabilities.
+
+However, MCP servers today are mostly operated locally without secure transport
+or authorization. OAuth authorization was recently added to the specification,
+but as of the time of writing, it is still new and not widely adopted.
+
+With Teleport's support for MCP servers, users can:
+- Access MCP servers on remote machines where transport is secured with TLS
+- Control access to specific MCP servers and specific tools defined by Teleport roles
+- Track activity through Teleport's audit log
+
+## Details
+
+To expedite the initial rollout, the first iteration of MCP server support will
+be implemented through App access.
+
+### UX
+
+#### editor - configure a filesystem MCP server and define RBAC
+To configure a [filesystem MCP
+server](https://github.com/modelcontextprotocol/servers/tree/main/src/filesystem)
+via Teleport app service:
+```yaml
+app_service:
+  enabled: true
+  apps:
+  - name: "dev-files"
+    description: "Shared files for developers"
+    labels:
+      env: "dev"
+    mcp:
+      command: "npx"
+      args: ["-y", "@modelcontextprotocol/server-filesystem", "/home/dev/files"]
```

Comment:
Running random binaries downloaded from the internet with root privileges makes me anxious 😅 

Perhaps it is too much for first version, but any kind of security hardening would be welcome. For example:
- strongly suggest using containers in our docs, examples
- make it possible to run commands with a different user

Reply:
@Tener @greedy52  does it have to be root and does teleport have to be responsible for spinning up the process?

Reply:
If we are spinning up the processes, I would assume we can leverage docker API or even interface with K8s to spin up jobs. I'm assuming this will be the most frequent use-case for isolation?